Fix annotations text tracking

atimmer opened the pull request:
## Description

This makes it possible to have an editor-only format that stays in
position when typing inside `RichText`.

Apply this to annotations, because it is necessary there.

Before this PR the annotations API worked via the functional `prepareEditableTree`. This works technically, but fails functionally. The rendering inside the TinyMCE instance is kept in sync with the annotations state. So when an annotation is from position 50 to position 100 the annotation will stay in those positions, this makes no sense from a user perspective. Instead what you want is that the annotation behaves like other formatting. It flows with edits of the text.

This morning I tried a first iteration by watching for edits to the `RichText` value and updating the annotation state accordingly, moving positions based on what has changed in the content. This was very unreliable and probably hard to make performant. While with this PR this tracking comes for free, because the DOM does it for us.

I'm thinking that I can iterate on this by having the annotations state be updated based on the formats the annotations module sees when running `valueToFormat`. Because that is the actual way that the annotation has moved in the text.

Note: With a fully functional paradigm I would update the annotations as edits from the user come in via events, but that's not feasible in the current `RichText` paradigm

## How has this been tested?

With the Yoast SEO 9.2-RC2 plugin and with the [annotations-tester plugin](https://github.com/atimmer/gutenberg-annotations-tester).

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

atimmer committed:
Allow filtering formatToValue and valueToFormat

This makes it possible to have an editor-only format that stays in
position when typing inside `RichText`.

Apply this to annotations, because it is necessary there.

atimmer commented:
@iseulde I've updated the description with more context.

iseulde commented:
With this example, the format stays in its place? Do you need a way to update the positions?

```js
wp.richText.registerFormatType( 'plugin/annotation', {
	title: 'invisible',
	tagName: 'mark',
	className: 'annotation',
	__experimentalGetPropsForEditableTreePreparation( select ) {
		return {
			isEnabled: select( 'core/edit-post' ).getActiveGeneralSidebarName() === 'edit-post/block',
		};
	},
	__experimentalCreatePrepareEditableTree( props ) {
		return ( formats, text ) => {
			if ( ! props.isEnabled ) {
				return formats;
			}

			const search = 'Gutenberg';
			const index = text.indexOf( search );

			if ( index === -1 ) {
				return formats;
			}

			const start = index;
			const end = index + search.length;

			const newValue = wp.richText.applyFormat( { text, formats }, { type: 'plugin/annotation' }, start, end );

			return newValue.formats;
		};
	},
} );
```

iseulde commented on packages/annotations/src/format/annotation.js:
```diff
@@ -64,19 +60,20 @@ export const annotation = {
-			return record.formats;
+			value = applyAnnotations( value, annotations );
+
+			return value;
```
Why do you need the ability to replace text?

atimmer replied:
I don't think I do, it's was just easier in rich-text to pass the whole value. I can easily change it to only allow changing the formats like `prepareEditableTree`.

iseulde replied:
Maybe I should have commented this. We don't want to allow `prepareEditableTree` to change anything other than the formats. See https://github.com/WordPress/gutenberg/pull/11595#discussion_r231578231

iseulde commented:
I would love a more elaborate description of what's happening in this PR.

> This makes it possible to have an editor-only format that stays in position when typing inside RichText.

The above test code proves that this is already possible.

iseulde commented on packages/editor/src/components/rich-text/index.js:
```diff
@@ -801,29 +804,87 @@ export class RichText extends Component {
 
+			// Allow primitives and arrays:
+			if ( ! isPlainObject( this.props[ name ] ) ) {
+				return this.props[ name ] !== prevProps[ name ];
```
Why are we comparing non `format_` props?

atimmer replied:
At this point in the code, is is guaranteed that `name` starts with `format_`.

iseulde replied:
I think this needs more commenting. I don't understand what's going on.

atimmer commented:
> I would love a more elaborate description of what's happening in this PR.
>
> > This makes it possible to have an editor-only format that stays in position when typing inside RichText.
>
> The above test code proves that this is already possible.

Pardon my inability to communicate this in text.

## What I want is:

### Before

Raw text:

```
Piece of text.
```

Annotated (classes left out for brevity):

```
Piece of <mark>text</mark>
```

Annotation state:

```js
{
 // ..
    range: {
        start: 9,
        end: 13,
    }
}
```

### After

Raw text:

```
Piece of texxt
```

Annotated:

```
Piece of <mark>texxt</mark>
```

Annotation state:

```js
{
 // ..
    range: {
        start: 9,
        end: 14,
    }
}
```

Given the fact that the annotation state only has positions to work with. I cannot use the method in the `__experimentalCreatePrepareEditableTree`, because I cannot search for a known string. 

*Note: I do understand that for marking specific strings the technique provided in your `invisible` example would work. An annotation doesn't mark a specific string though, it references a selection of text, that can grow or shrink or move.*

youknowriad commented on packages/annotations/src/format/annotation.js:
```diff
@@ -64,19 +60,20 @@ export const annotation = {
-		return {
-			annotations: select( 'core/annotations' ).__experimentalGetAnnotationsForRichText( blockClientId, richTextIdentifier ),
-		};
+		return select( 'core/annotations' ).__experimentalGetAnnotationsForRichText( blockClientId, richTextIdentifier );
```
Is this wrong? the method is called `getProps` I expected it to return an object here?

atimmer replied:
The thing I tried to solve here was not re-rendering unless necessary. `select( 'core/annotations' ).__experimentalGetAnnotationsForRichText( blockClientId, richTextIdentifier )` is cached based on the annotations for a block, so it will only re-render if a different annotation is added. 

With:

```js
return {
	annotations: select( 'core/annotations' ).__experimentalGetAnnotationsForRichText( blockClientId, richTextIdentifier ),
};

it would create a new object every time the function was called, so it would trigger a re-render. And I didn't want to introduce another memoized location, when a simpler solution was available.

youknowriad replied:
My hope/thoughts, was that `withSelect` already checks the props (using shallow equality) and we can probably use the same mechanism to do the same.

youknowriad commented:
So I understand properly:

```
__experimentalCreatePrepareEditableTree
```

was solving the Annotations => RichText (link)


but we need an `onChange` somewhere and this is what `__experimentalCreateValueToFormat` is about right?

---
I think if we assume that the RichText component is capable of editing the format ranges. (which it is obviously doing since we have "boundaries" support), it's legitimate to expect the format to receive the changes (if it wants to do something with it).

Still not sure about the implementation details, I need to think a little bit more but I agree with the issue we're trying to solve being legitimate.

atimmer commented:
> but we need an onChange somewhere and this is what __experimentalCreateValueToFormat is about right?

Yes, that's currently not in this PR. But my thinking is that I can update the annotation state based on the `value` in `__experimentalCreateValueToFormat`.

> I think if we assume that the RichText component is capable of editing the format ranges. (which it is obviously doing since we have "boundaries" support), it's legitimate to expect the format to receive the changes (if it wants to do something with it).

That would work too.

> Still not sure about the implementation details, I need to think a little bit more but I agree with the issue we're trying to solve being legitimate.

I wanted to at least put something up that works. I'm open to better solutions.

iseulde commented:
So what you currently have is: annotations state => value formats.
What you want as well is: value formats => annotations state.

So if we call same function with the value formats whenever they change, you can derive new annotation state from that? This probably requires you to pass some id to the format so you can identify it later on.

I'm not sure how this PR does that. The lack of test doesn't make it clearer.

youknowriad commented:
What about this API?

```js
const format = {
   prepareEditableValue: select => (value, context) => { return formats; }
    onChangeEditableValue: dispatch => (formats, context) => { // do something }
}
```

I feel this is simple enough to be understood properly it assumes we select props when rendering and we dispatch things onChange.

The downside here is that we can't "optimize" `prepareEditableValue` in the Framework, it's left to the format author. Concretly, it means the format author would need to `memoize` the  `prepareEditableValue` inner function it self to avoid generating new formats if the props (computed using the `select` argument) didn't change. 

Thoughts?

youknowriad commented:
Alternatively, if we were to keep the separation between "receiving the props" and "providing the formats based on the props and the value" to be able to bundle the performance optimization into the framework, I'll do the same for the opposite direction:


```js
const format = {
   getPropsForEditableTreePreparation => (select, context) => props,
   prepareEditableValue: (value, props, context) => { return formats; },
   
    getPropsForEditableTreeChangeHandler => (dispatch, context) => props,
    onChangeEditableValue: (formats, props, context) => { // do something }
}
```

atimmer committed:
Update annotations based on changes in `RichText`

When the user types before or in an annotation we want the annotation
to be adjusted accordingly. We can determine the location of the mark
elements in the `RichText` based on the value passed to
`__experimentalCreateValueToFormat` so we can update the annotations
based on the values expressed there.

atimmer commented:
> So what you currently have is: annotations state => value formats.
> What you want as well is: value formats => annotations state.

I think that is correct.

> So if we call same function with the value formats whenever they change, you can derive new annotation state from that? This probably requires you to pass some id to the format so you can identify it later on.

Yes, I've implemented this in the latest commit: https://github.com/WordPress/gutenberg/pull/11861/commits/ea44b15902a88b3e15f4d2744dcf43975982262b.

> I'm not sure how this PR does that. The lack of test doesn't make it clearer.

It didn't yet update the annotations state because I hadn't gotten to that yet. In the last commit I've implemented that part.

atimmer commented:
> The downside here is that we can't "optimize" prepareEditableValue in the Framework, it's left to the format author. Concretly, it means the format author would need to memoize the prepareEditableValue inner function it self to avoid generating new formats if the props (computed using the select argument) didn't change.
> 
> Thoughts?

I think we cannot leave this hole open. If all `RichText` instances re-render on any change the performance is instantly completely broken.

> ```js
> const format = {
>    getPropsForEditableTreePreparation => (select, context) => props,
>    prepareEditableValue: (value, props, context) => { return formats; },
>    
>     getPropsForEditableTreeChangeHandler => (dispatch, context) => props,
>     onChangeEditableValue: (formats, props, context) => { // do something }
> }
> ```

What is the benefit of the `getPropsForEditableTreeChangeHandler` handler? Is there a huge downside if implementation call `wp.dispatch( ... )` inside `onChangeEditableValue`?

@youknowriad The version I have now works completely, I will try to implement your proposed second solution to see if that works. I don't know yet how I am going to work around/with https://github.com/WordPress/gutenberg/blob/master/packages/rich-text/src/create.js#L60, but I'll try to find a solution.

youknowriad commented:
> What is the benefit of the getPropsForEditableTreeChangeHandler handler? Is there a huge downside if implementation call wp.dispatch( ... ) inside onChangeEditableValue?

I think we discussed this in a previous #core-js but I agree that it's not obvious :). Ideally we should never use `wp.dispatch` or `wp.select` (the globals) in Core Packages. We have an issue to move away from these singletons.

The bug that can happen here is that you'd rely on the wrong registry. For example if you have a Gutenberg editor inside another Gutenberg editor (which can happen for reusable blocks), you won't be calling the right registry if you do `wp.dispatch` since the registry will be overriden using a `RegistryProvider`.

atimmer committed:
Annotations: Refactor RichText APIs

Instead of having APIs for `formatToValue` and `valueToFormat`, create
an `__experimentalCreateOnChangeEditableValue` function that is called
when the editable value changes. The annotations API can use this to
update its state.

youknowriad commented:
In my mind it's slightly different (there's two outside words) and TinyMCE is not really important no matter the approach we choose.

Also, pasting remove the annotaitons => This doesn't seem like a bug for me. When you paste something, you are overwriting the whole value?

Here's how I see it.

![untitled diagram 1](https://user-images.githubusercontent.com/272444/48632314-e59c0480-e9c0-11e8-909f-2a967a496f64.png)

iseulde commented:
I think `onChangeEditableValue: (formats, props, context) => {}` is closer to what we want. With `prepareEditableValue` you're deriving new formats from annotation state. With `onChangeEditableValue` you'd derive new annotation state from the formats (likely in a different position). So you need to loop through all formats and set new start and end positions. We could provide helper functions if this would make it easier, e.g. `getFormatPositions( value ) = [ { format, start, end }, ... ]`. I'll have a look on Monday to implement it if needed.

youknowriad commented:
Unit tests are failing. Probably docs needs to be regenerated or something.

I'd love if we can update the annotations e2e test to check updates, say like we type in an annotation and we check that it's being updated properly in the state or something.

ellatrix reviewed (changes requested):
I don't see any tests at all. This would make it hard for anyone changing RichText in the future to know if they broke the API.

iseulde commented on packages/annotations/src/format/annotation.js:
```diff
@@ -52,31 +57,104 @@ export function removeAnnotations( record ) {
+	const positions = {};
+
+	formats.forEach( ( characterFormats, i ) => {
+		characterFormats = characterFormats || [];
```
This should always be an array, so the casting seems unnecessary?

atimmer replied:
It can be `undefined` right, for when no formatting is present on that character?

iseulde replied:
It's a sparse array, there's no indices set as undefined, they're just not there. `forEach` will only loop over any indices that are set, which is only arrays.

iseulde replied:
Try:

```js
[ , [] ].forEach( ( { forEach } ) => {} );
[ undefined, [] ].forEach( ( { forEach } ) => {} );
```

iseulde commented on packages/annotations/src/format/annotation.js:
```diff
@@ -52,31 +57,104 @@ export function removeAnnotations( record ) {
+
+	formats.forEach( ( characterFormats, i ) => {
+		characterFormats = characterFormats || [];
+		characterFormats = characterFormats.filter( ( format ) => format.type === FORMAT_NAME );
```
Could these two loops be merged?

atimmer commented:
> Also, pasting remove the annotaitons => This doesn't seem like a bug for me. When you paste something, you are overwriting the whole value?

When pasting, I think the annotations should stay in place if the text they apply to is still present. If I cut and paste a sentence from the middle of the paragraph to the beginning of the paragraph the annotations should probably stay in place.

iseulde commented on packages/annotations/src/format/annotation.js:
```diff
@@ -52,31 +57,104 @@ export function removeAnnotations( record ) {
+			const positions = retrieveAnnotationPositions( formats );
+			const { removeAnnotation, updateAnnotationRange, annotations } = props;
+
+			updateAnnotationsWithPositions( annotations, positions, { removeAnnotation, updateAnnotationRange } );
```
Why can `retrieveAnnotationPositions` and `updateAnnotationsWithPositions` not be done in one go, something like `updateAnnotationsFromFormats`?

atimmer replied:
I felt like these two functions were good to make the code more understandable, they can be rolled into one if that's better.

atimmer commented on packages/editor/src/components/rich-text/index.js:
```diff
@@ -444,6 +448,18 @@ export class RichText extends Component {
+	 * @param {Array}  formats The formats of the latest rich-text value.
+	 * @param {string} text    The text of the latest rich-text value.
+	 */
+	onChangeEditableValue( { formats, text } ) {
```
I don't kwow if I understand the `rich-text` package enough to be able to rewrite this code myself. Should I try to take a stab at writing those tests so you can verify your code?

iseulde replied:
That sounds good!

atimmer committed:
Fix and add annotations reducer tests

atimmer committed:
Move annotations file to separate file.

atimmer committed:
Add e2e tests for annotation text tracking

atimmer committed:
Change e2e test annotations sidebar name to be unique

atimmer committed:
Update docs

mcsf commented on packages/annotations/src/store/selectors.js:
```diff
@@ -61,5 +65,7 @@ export const __experimentalGetAnnotationsForRichText = createSelector(
-	return state.all;
+	return flatMap( state, ( annotations ) => {
+		return annotations;
+	} );
```
This is equivalent to `flatMap( state, identity )`, which can be reduced to `flatten( state )`, unless I misread.

iseulde commented on test/e2e/specs/__snapshots__/plugins-api.test.js.snap:
```diff
@@ -1,3 +1,3 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
-exports[`Using Plugins API Sidebar Should open plugins sidebar using More Menu item and render content 1`] = `"<div class=\\"components-panel__header edit-post-sidebar-header__small\\"><span class=\\"edit-post-sidebar-header__title\\">(no title)</span><button type=\\"button\\" aria-label=\\"Close plugin\\" class=\\"components-button components-icon-button\\"><svg aria-hidden=\\"true\\" role=\\"im ...
```
What is this?

youknowriad replied:
I think the difference here is that it was moved to a separate test. there's no annotations button anymore.

atimmer replied:
Yes, that is the reason.

youknowriad reviewed (approved):
I'm approving this as it seems in a decent shape to include in 4.5 but let's make sure we follow up on the comments.
